session.go

```diff
@@ -312,16 +314,27 @@ func (s *Session) Ping() (dur time.Duration, err error) {
 		s.pingLock.Unlock()
 	}()
 
-	// Send the ping request
-	hdr := encode(typePing, flagSYN, 0, activePing.id)
-	if err := s.sendMsg(hdr, nil, nil); err != nil {
-		return 0, err
+	// Send the ping request, waiting at most one connection write timeout
+	// to flush it.
+	timer := time.NewTimer(s.config.ConnectionWriteTimeout)
+	defer timer.Stop()
+	select {
+	case s.pingCh <- activePing.id:
+	case <-timer.C:
+		return 0, ErrTimeout
+	case <-s.shutdownCh:
+		return 0, s.shutdownErr
+	}
+
+	// Wait for a response, again waiting at most one write timeout.
```

Comment:
this comment is out of place. the waiting for a response happens after cleaning up the timer.

Reply:
I originally put the comment above the timer logic to mirror the timer logic above the previous section. I've rearranged things to improve the grouping.